src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs

```diff
@@ -406,6 +413,35 @@ private static (RoutePattern resolvedRoutePattern, IDictionary<string, string?>
         {
             perRouteConventions[i](builder);
         }
+
+        if (builder is RouteEndpointBuilder { RouteHandlerFilterFactories.Count: > 0 } reb && action is ControllerActionDescriptor cad)
+        {
+            var routeHandlerFilters = reb.RouteHandlerFilterFactories;
+
+            // Make them null to avoid the default behavior
+            reb.RouteHandlerFilterFactories = null;
+
+            var context = new RouteHandlerContext(cad.MethodInfo, builder.Metadata, _serviceProvider);
+
+            // Most inner delegate needs a ControllerRouteHandlerInvocationContext to function
+            RouteHandlerFilterDelegate del = static async (context) =>
+            {
+                var controllerRhiContext = (ControllerRouteHandlerInvocationContext)context;
```

Comment:
@captainsafia unlike the RequestDelegateFactory, this needs to be of a specific type which makes me think we should fail if you replace the instance since it won't work in all cases.